book/docs/Improving_science_code.md

```diff
@@ -0,0 +1,75 @@
+
+<!-- Potentially a drawing of turles all the way down. -->
+
+It's not uncommon for code to be structured in that way and often it has to do with a desire for code to be extremely DRY (Do not Repeat Yourself) or to have a consistent level of abstraction at every level. Both of these concepts are good in theory, but in practice they can make code hard to debug. 
+
+Here are some questions to ask when you are writing a function:
+
+- Is it going to be used more than three times?
+- Does it get different inputs each time it is used?
+- Does it make the code easier to understand?
+- Is the function more than 4 lines long?
+
+If one or two of these are "no" then consider whether a function is really needed.
+
+## Code Contents
+
+### Naming
+You can make any code more legible by using variable names that are clear and descriptive. This means using full words and substituting underscores for spaces.
+
+| Do | Don't |
+| -- | ----- |
+| `region_name = "USA"` | `regnm = "USA"` |
+
+
+### Constants
+Move science parameters into one file and import them directly from there in the places where you need them. Constants should not be passed as paramenters to functions. Consider using a library like [pydantic-settings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/) to make it easy to override constants.
+
+### Making code faster
+Writing and reading code take time just the same way that running code takes time, so if a function works and runs for 10 minutes once a year then maybe it doesn't need to be faster. Another way of putting it is: the more often a piece of code runs, the more valuable it is to make it run fast.
+
+If you have decided that it is worthwile to speed up your code the first thing to do is look for for-loops. Go to the deepest nested for-loop and see what happens in there. If it is math, see whether you can use vectors (for instance numpy arrays) rather than iterating. If there are function calls, think about what the inputs to the functions will be. Ideally you should only run a funciton if there are new inputs. That includes IO so if you are repeatedly reading the same file within a for-loop, consider moving the read up and out.
+
+If you aren't sure which approach will be faster create a minimal example (or even better a test) and try it both ways and measure how long each takes.
+
+## Code Output
+
+When writing output make sure to use stable and popular file formats. Here are some examples:
+ 
+ - for dictionary-like data: json
+ - for tables: parquet
+ - for rasters: tiff or COG
+ - for nd arrays: zarr
+
+When writing output you often have control over how the data gets compressed and how each datatype is represented (think 8 bit vs 64 bit). You probably don't need to think about that when you first write a file, but if you are trying to make reads and writes more performant, that is something to consider. 
+
+### The case against Pickles
```

Comment:
Make it clear at the top this is a Python particular example, but similar issues exist in other options. e.g. RData vs RDS

Reply:
Maybe also be clear, saving objects is hard - better to save to file formats that can be read back as needed, and more generally usable for cross compatibility.